src/compiler/core.ts

```diff
@@ -2035,11 +2035,22 @@ namespace ts {
      * Takes a string like "jquery-min.4.2.3" and returns "jquery"
      */
     export function removeMinAndVersionNumbers(fileName: string) {
+        // We used to use the regex /[.-]((min)|(\d+(\.\d+)*))$/ and would just .replace it twice.
+        // Unfortunately, that regex has O(n^2) performance because v8 doesn't match from the end of the string.
+        // Instead, we now split the input on the . and - characters, and drop the trailing segements
+        // which match the /^[.-](min|\d+)$/ pattern, which should be O(n). (This technically allows more `.min` repetitions than before, but that's fine)
+
         // Match a "." or "-" followed by a version number or 'min' at the end of the name
-        const trailingMinOrVersion = /[.-]((min)|(\d+(\.\d+)*))$/;
+        const trailingMinOrVersion = /^[.-](min|\d+)$/;
 
-        // The "min" or version may both be present, in either order, so try applying the above twice.
-        return fileName.replace(trailingMinOrVersion, "").replace(trailingMinOrVersion, "");
+        // Split on . and -, then remove all trailing segments that match
+        const parts = fileName.split(/(?=[.-])/g);

```

Comment:
.split and .join are generally quite expensive, should be faster still to just use a few `lastIndexOf` calls for '.' and '-' and trim the string if the bit after conforms

Reply:
Microbenchmarks say this currently is much slower (about 5x) the original implementation. Do you have examples of sample input this will perform poorly on?

Reply:
Try this:
```ts
export function removeMinAndVersionNumbers(fileName: string) {
    let end: number = fileName.length;

    for (let pos = end - 1; pos > 0; pos--) {
        let ch: number = fileName.charCodeAt(pos);
        if (ch >= CharacterCodes._0 && ch <= CharacterCodes._9) {
            // Match a \d+ segment
            do {
                --pos;
                ch = fileName.charCodeAt(pos);
            } while (pos > 0 && ch >= CharacterCodes._0 && ch <= CharacterCodes._9);
        } else if (pos > 4 && (ch === CharacterCodes.n || ch === CharacterCodes.N)) {
            // Looking for "min" or "min"
            // Already matched the 'n'
            --pos;
            ch = fileName.charCodeAt(pos);
            if (ch !== CharacterCodes.i && ch !== CharacterCodes.I) {
                break;
            }
            --pos;
            ch = fileName.charCodeAt(pos);
            if (ch !== CharacterCodes.m && ch !== CharacterCodes.M) {
                break;
            }
            --pos;
            ch = fileName.charCodeAt(pos);
        } else {
            // This character is not part of either suffix pattern
            break;
        }

        if (ch !== CharacterCodes.minus && ch !== CharacterCodes.dot) {
            break;
        }
        
        end = pos;
    }

    // end might be fileName.length, in which case this should internally no-op
    return end === fileName.length ? fileName : fileName.slice(0, end);
}
```

Reply:
It probably does not, hence the quick and dirty original solution, but since you took the time to write the reverse scanner for the function, I've included it. ❤️